### ISSUE TITLE
lots of time wasted on `count_deltas()`

### ISSUE THREAD

**Issue by problame** [neondatabase/neon#6861]
# Problem

original thread: https://neondb.slack.com/archives/C033RQ5SPDH/p1708513450565049

Now that [the flamegraphs are fixed](https://github.com/neondatabase/neon/pull/6764), I took one on ps-2 ap-southeast-1 to investigate the elevanted CPU usage after enabling tokio-epoll-uring there.
That investigation isn't the subject of this thread though, but, the general finding of where that PS is spending its time.
`LayerMap::count_deltas` inside time_for_new_image_layer completely dominates the CPU usage there.
AFAICT that is called for every tenant, even if the layer map hasn't changed.

This is  wasteful.

# Solution

If the layer map and partitioning is the same as in an earlier call, early-exit in `time_for_new_image_layer` to avoid the call to `count_deltas()`.

# Tasks

```[tasklist]
### Tasks
- [ ] Add a draft title or issue reference here
```

**Comment by hlinnaka** [comment]
The new compaction code in https://github.com/neondatabase/neon/pull/6830/ no longer calls `count_deltas`. (It needs testing to see if it introduces other problems of course)

**Comment by problame** [comment]
Yeah, aware, @arpad-m  is going to work on compaction, but, it'll be many more weeks until it lands, I think.

**Comment by VladLazar** [comment]
Update:
* WAL based solution merged and will be released this week (2024-04-02)
* Need to check things have improved post release. Keeping the ticket open until then.
